main.go

```diff
@@ -1,13 +1,71 @@
 package main
 
 import (
-	"log"
+	"context"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
 
+	"github.com/netlify/gotrue/api"
 	"github.com/netlify/gotrue/cmd"
+	"github.com/netlify/gotrue/observability"
+	"github.com/sirupsen/logrus"
 )
 
+func init() {
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+}
+
 func main() {
-	if err := cmd.RootCommand().Execute(); err != nil {
-		log.Fatal(err)
+	execCtx, execCancel := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGHUP, syscall.SIGINT)
+	defer execCancel()
+
+	go func() {
+		<-execCtx.Done()
+		logrus.Info("received graceful shutdown signal")
+	}()
+
+	// command is expected to obey the cancellation signal on execCtx and
+	// block while it is running
+	if err := cmd.RootCommand().ExecuteContext(execCtx); err != nil {
+		logrus.WithError(err).Fatal(err)
+	}
+
+	shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), time.Minute)
+	defer shutdownCancel()
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// wait for API servers to shut down gracefully
+		api.WaitForCleanup(shutdownCtx)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// wait for metrics and trace exporters to shut down gracefully
+		observability.WaitForCleanup(shutdownCtx)
+	}()
```

Comment:
hmm i don't really get why we need to do `WaitForCleanup(ctx)` twice here - also, do we need to have the same function repeated in 2 packages? afaik, there shouldn't be a cyclical dependency and we can have `WaitForCleanup` in either the api or observability package

Reply:
Well one waits for cleanups scoped in the `api` package, and the other for cleanups scoped in the `observability` package. An alternative approach is to have a yet third package, for example `cleanup` which could somewhat dangerously require exposing internal package state to it, so that waiting for cleanup be made possible.

What alternative are you proposing?

Reply:
Also there is no cyclical dependency, that's yet another reason why two separate functions are a better idea. `api.WaitForCleanup` waits for started API servers and other cleanup jobs (like, for example that feature about cleaning up stale refresh tokens), while `observability.WaitForCleanup` waits for metrics and traces to be properly sent before the process dies. Two completely different things are waited upon...

Reply:
ah i see, nah i don't really have an alternative in mind for now but it just came across as slightly strange to have the same function duplicate across 2 different packages